src/seraphis_mocks/mock_http_client_pool.h

```diff
@@ -0,0 +1,139 @@
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// NOT FOR PRODUCTION
+// TODO for production-ready:
+// 1. Implement close_connections.
+// 2. Require the pool respect max_connections.
+
+#pragma once
+
+//local headers
+#include "misc_language.h"
+#include "net/http.h"
+#include "net/http_client.h"
+#include "rpc/core_rpc_server_commands_defs.h"
+#include "storages/http_abstract_invoke.h"
+
+//third party headers
+
+//standard headers
+#include <mutex>
+#include <string>
+#include <vector>
+
+//forward declarations
+
+namespace sp
+{
+namespace mocks
+{
+////
+// ClientConnectionPool
+// - wraps a pool of network client connections to enable concurrent requests
+///
+class ClientConnectionPool
+{
+public:
+//constructors
+    ClientConnectionPool(
+        const std::string &daemon_address,
+        const boost::optional<epee::net_utils::http::login> daemon_login = boost::none,
+        const epee::net_utils::ssl_options_t ssl_support = epee::net_utils::ssl_support_t::e_ssl_support_autodetect,
+        const std::string &proxy = "",
+        const std::size_t max_connections = 20):
+            m_daemon_address{daemon_address},
+            m_daemon_login{daemon_login},
+            m_ssl_support(ssl_support),
+            m_proxy{proxy},
+            m_max_connections{max_connections}
+    {
+        m_http_client_pool.reserve(max_connections);
+    }
+
+//member functions
+    enum http_mode { JSON, BIN, JSON_RPC };
+
+    /// Use an http client from the pool to make an RPC request to the daemon
+    template <typename COMMAND_TYPE>
+    bool rpc_command(const http_mode &mode,
+        const std::string &command_name,
+        const typename COMMAND_TYPE::request &req,
+        typename COMMAND_TYPE::response &res)
+    {
+        // Acquire an http client from the connection pool
+        std::size_t http_client_index = acquire_unused_http_client();
```

Comment:
```suggestion
        std::size_t http_client_index = this->acquire_unused_http_client();
```
Disambiguate (same as `self.` in rust).